# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: devtools/server/startup/worker.js
# Commit: 895cf08432c4
# Full Hash: 895cf08432c40fe371ff83e846cb161430c6be4b
# Author: Nicolas Chevobbe <nchevobbe@mozilla.com>
# Date: 2020-10-27 04:41:26
# Regressor Bug: 1633712
# File Overlap Count: 1
# Description:
#   Bug 1633712 - [devtools] Create WorkerTargets as soon as possible. r=ochameau,devtools-backward-compat-reviewers.
#   
#   This patch adds support for dedicated worker targets in the Watcher actor.
#   Shared and Service workers are not handled yet.
#   
# ==============================================================================

diff -r 2ce79a9ed2d5 -r 895cf08432c4 devtools/server/startup/worker.js
--- a/devtools/server/startup/worker.js	Mon Oct 26 12:52:26 2020 +0000
+++ b/devtools/server/startup/worker.js	Mon Oct 26 12:53:43 2020 +0000
@@ -54,7 +54,7 @@
 // that, we handle a Map of the different connections, keyed by forwarding prefix.
 const connections = new Map();
 
-this.addEventListener("message", function(event) {
+this.addEventListener("message", async function(event) {
   const packet = JSON.parse(event.data);
   switch (packet.type) {
     case "connect":
@@ -64,7 +64,11 @@
       const connection = DevToolsServer.connectToParent(forwardingPrefix, this);
 
       // Step 4: Create a WorkerTarget actor.
-      const workerTargetActor = new WorkerTargetActor(connection, global);
+      const workerTargetActor = new WorkerTargetActor(
+        connection,
+        global,
+        packet.workerDebuggerData
+      );
       // Make the worker manage itself so it is put in a Pool and assigned an actorID.
       workerTargetActor.manage(workerTargetActor);
 
@@ -80,6 +84,7 @@
       // it that a connection has been established.
       connections.set(forwardingPrefix, {
         connection,
+        workerTargetActor,
       });
 
       postMessage(
@@ -90,12 +95,44 @@
         })
       );
 
+      // We might receive data to watch.
+      if (packet.options?.watchedData) {
+        const promises = [];
+        for (const [type, entries] of Object.entries(
+          packet.options.watchedData
+        )) {
+          promises.push(workerTargetActor.addWatcherDataEntry(type, entries));
+        }
+        await Promise.all(promises);
+      }
+
+      break;
+
+    case "add-watcher-data-entry":
+      await connections
+        .get(packet.forwardingPrefix)
+        .workerTargetActor.addWatcherDataEntry(
+          packet.dataEntryType,
+          packet.entries
+        );
+      postMessage(JSON.stringify({ type: "watcher-data-entry-added" }));
+      break;
+
+    case "remove-watcher-data-entry":
+      await connections
+        .get(packet.forwardingPrefix)
+        .workerTargetActor.removeWatcherDataEntry(
+          packet.dataEntryType,
+          packet.entries
+        );
       break;
 
     case "disconnect":
       // This will destroy the associate WorkerTargetActor (and the actors it manages).
-      connections.get(packet.forwardingPrefix).connection.close();
-      connections.delete(packet.forwardingPrefix);
+      if (connections.has(packet.forwardingPrefix)) {
+        connections.get(packet.forwardingPrefix).connection.close();
+        connections.delete(packet.forwardingPrefix);
+      }
       break;
 
     case "rpc":